File: src/main/scala/ge/dm/year2022/RucksackReorganization.scala
package ge.dm.year2022

import ge.dm.utils.Reader

object RucksackReorganization {

  def main(args: Array[String]): Unit = {
    print(solve())
  }

  def solve(): Int = {
    val data = Reader.readFileString("src/resources/year2022/Day3RucksackReorganization.txt")

    data.indices.map( x => checkRuckSack(data(x))).sum
  }

  def checkRuckSack(items: String): Int = {
    val s: Set[Char] = items.substring(0, items.length / 2).toSet
    items.substring(items.length / 2, items.length)
      .filter(i => s.contains(i))
      .take(1)
      .map(i => calculateScore(i)).toList.head
  }

  def calculateScore(ch: Char): Int = {
    if (ch.isLower) {
      ch - 96
    } else {
      ch - 38
    }
  }


}

File: src/main/scala/ge/dm/year2022/java/RucksackReorganization.java
package ge.dm.year2022.java;

import ge.dm.utils.JReader;

import java.util.stream.Collectors;

public class RucksackReorganization {

    public static void main(String[] args) {
        System.out.println(solve());
    }

    public static int solve() {
        var lines = JReader.ReadFileAsString("src/resources/year2022/Day3RucksackReorganization.txt");

        return lines.stream()
                .map(RucksackReorganization::checkRuckSack)
                .reduce(0, Integer::sum);
    }

    public static int checkRuckSack(String items) {
        var set = items.substring(0, items.length()/2)
                .chars()
                .mapToObj(i -> (char) i)
                .collect(Collectors.toSet());
        return items.substring(items.length()/2).chars().mapToObj((i) -> (char) i)
                .filter(set::contains)
                .findFirst().map(RucksackReorganization::calculateScore).get();
    }

    public static int calculateScore(char ch) {
        if (Character.isLowerCase(ch)) {
            return ch - 96;
        } else {
            return ch - 38;
        }
    }

}
